mass_mailing_list_dynamic/models/mail_mass_mailing_list.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Jairo Llopis <jairo.llopis@tecnativa.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+from odoo.tools import safe_eval
+
+
+class MassMailingList(models.Model):
+    _inherit = "mail.mass_mailing.list"
+
+    sync_method = fields.Selection(
+        [
+            ("add", "Only add new records"),
+            ("full", "Add and remove records as needed"),
+        ],
+        help="Choose the syncronization method for this list if you want to "
+             "make it dynamic",
+    )
+    sync_domain = fields.Char(
+        default="[('opt_out', '=', False), ('email', '!=', False)]",
+        help="Filter partners to sync in this list",
+    )
+
+    def action_sync(self):
+        """Sync contacts in dynamic lists."""
+        Contact = self.env["mail.mass_mailing.contact"]
+        Partner = self.env["res.partner"]
+        # Skip non-dynamic lists
+        dynamic = self.filtered("sync_method")
+        for one in dynamic:
+            sync_domain = safe_eval(one.sync_domain) + [("email", "!=", False)]
+            desired_partners = Partner.search(sync_domain)
+            current_contacts = Contact.search([("list_id", "=", one.id)])
+            current_partners = current_contacts.mapped("partner_id")
+            # Remove undesired contacts when synchronization is full
+            if one.sync_method == "full":
+                current_contacts.search([
+                    ("partner_id", "not in", desired_partners.ids),
+                ]).unlink()
+            # Add new contacts
+            for partner in desired_partners - current_partners:
+                Contact.create({
+                    "list_id": one.id,
+                    "partner_id": partner.id,
+                })
+        # Invalidate cached contact count
+        self.invalidate_cache(["contact_nbr"], dynamic.ids)
+
+    # TODO Delete when fixed https://github.com/OCA/oca-decorators/issues/6
+    def write(self, vals):
```

Comment:
I don't think we should sync on create/write. Just when manually done or when sending mailings.